content/shmem_malloc_hints.tex

```diff
@@ -0,0 +1,97 @@
+    
+    In addition to the \VAR{size} argument, the \VAR{hint} argument is provided by the user. 
+    The \VAR{hint} describes the expected manner in which the \openshmem program may use the allocated memory.
+    The valid usage hints are described in Table~\ref{usagehints}. Multiple hints are expressed as \CONST{OR} of \VAR{hints}.
+    
+    The information provided by the \VAR{hint} is used to optimize for performance by the implementation. 
+    If the implementation cannot optimize, the behavior is same as \FUNC{shmem\_malloc}.
+    If more than one hint is provided, the implementation will make the best effort to use one or more hints 
+    to optimize performance. 
+            
+    The \FUNC{shmem\_malloc\_with\_hint} routine is provided  so that multiple \acp{PE} in a program can allocate symmetric,
+    remotely accessible memory blocks.  When no action is performed, these
+    routines return without performing a barrier. Otherwise, the routine will call a barrier on exit.
+    This ensures that all \acp{PE} participate in the memory allocation, and that the memory on other
+    \acp{PE} can be used as soon as the local \ac{PE} returns. The implicit barrier performed by this routine will quiet the
+    default context.  It is the user's responsibility to ensure that no communication operations involving the given memory block are pending on
+    other contexts prior to calling the \FUNC{shmem\_free} and \FUNC{shmem\_realloc} routines.
+    The user is also responsible for calling these routines with identical argument(s) on all
+    \acp{PE}; if differing \VAR{size}, or \VAR{hint} arguments are used, the behavior of the call
+    and any subsequent \openshmem calls is undefined.
+}
+
+\apireturnvalues{
+    The \FUNC{shmem\_malloc\_with\_hint} routine returns a pointer to the allocated space;
+    otherwise, it returns a null pointer.
+}
+
+\apinotes{ 
+}
+
+\apiimpnotes{
+}
+	\begin{longtable}{|p{0.45\textwidth}|p{0.5\textwidth}|}
+	\hline
+	\textbf{Hints} & \textbf{Usage hint}
+	\tabularnewline \hline
+	\endhead
+	%%
+	\LibConstDecl{SHMEM\_HINT\_NONE} &
+	Behavior same as shmem\_malloc
+	\tabularnewline \hline
+		
+	\LibConstDecl{SHMEM\_HINT\_LOW\_LAT\_MEM} &
+	Allocate memory on low-latency storage	
+	\tabularnewline \hline
+	
+	\LibConstDecl{SHMEM\_HINT\_HIGH\_BW\_MEM} &
+	Allocate memory on high-bandwidth storage	
+	\tabularnewline \hline
+
+	\LibConstDecl{SHMEM\_HINT\_PSYNC} &
+	Memory used as \CONST{PSYNC} array
+	\tabularnewline \hline
+		
+	\LibConstDecl{SHMEM\_HINT\_PWORK} &
+	Memory used as \CONST{PWORK} array
+	\tabularnewline \hline
+	
+	\LibConstDecl{SHMEM\_HINT\_ATOMICS} &
+	Memory used for \VAR{Atomic} operations
+	\tabularnewline \hline
+	
+	\LibConstDecl{SHMEM\_HINT\_SIGNAL} &
+	Memory used for \VAR{signal} operations
+	\tabularnewline \hline
+
+	\TableCaptionRef{Memory usage hints}
+        \label{usagehints}
+    \end{longtable}
```

Comment:
Also we should to state that neither alignment requirements or memory properties, such as cache line size are not get impacted by the hint. We also shell state the memory semantics for local assess (load and store) are not impacted by the hint.

Reply:
AFAIU, for the shmem_malloc routine the implementation is free to allocate the memory, which is either cache aligned or not. One of the constraints is that it should be word-aligned. Similarly, the memory access model (which is yet be defined or clarified here https://github.com/openshmem-org/specification/issues/229) will provide certain access guarantees to the memory allocated by shmem_malloc. In both cases, the proposal intends to follow the semantics of shmem_malloc.

Reply:
The memory gets eventually mapped into the core. The core architecture defined multiple way how it can be mapped. Each on of the mapping has own semantics and constrains. For example semantics between normal cacheable (write back) and non-cacheable (WC) is very different. Since user has direct access through the pointer, code that worked on one machine will break on another with exception or even data corruption.

Reply:
In order to support something like this, you have to remove shmem_ptr and prohibit any direct asses to the memory through load and store semantics. Next you would have to introduce shmem_memcpy function to copy-in-out shmem_malloced region.